proposals/lip-introduce-events-and-add-events-root-to-block-headers.md

```diff
@@ -0,0 +1,264 @@
+An event emitted during transaction processing cannot have any of these reserved constant in the `topics` array.
+
+#### `index`
+
+The event `index` is an integer between `0` and `MAX_EVENTS_PER_BLOCK - 1`, set to the number of events that were previously emitted in the same block. 
+As a consequence, during a single block, at most `MAX_EVENTS_PER_BLOCK` events can be emitted.
+
+### Error Handling and Events
+
+For events emitted during the command execution, modules specify whether the event is reverted or not in case of error (i.e., whether it is included in the events tree or not). 
+In all other application domain stages errors cannot occur in a valid block (thus, also state changes are never reverted) so that it is not necessary to specify if the event is revertible.
+
+
+### Events Tree
+
+Events are inserted into the events tree once for each entry in the event `topics` array. 
+The leaf key is the concatenation of the hash of the element in the topics array, the event `index` converted to a binary string of length `EVENT_INDEX_LENGTH_BITS` and concatenated with the topic index of `TOPIC_INDEX_LENGTH_BITS` bits (corresponding to the entry in the `topics` array), for a total of `TOTAL_INDEX_LENGTH_BYTES` bytes. 
+The leaf value is the serialization of the event properties using the `eventSchema` given above. 
+
+In summary, for each event one key-value pair is added to the events tree for each entry `topic` in the associated `topics` array, where:
+
+* `key = trunc_hash(topic) || uint_to_bytes((event index << TOPIC_INDEX_LENGTH_BITS) + topic index)`
+* `value` = serialization of event according to `eventSchema`,
+
+where `trunc_hash` is the `SHA-256` hash function truncated to the first `TOPIC_HASH_LENGTH_BYTES` bytes and the `uint_to_bytes` function serializes an integer as a big-endian bytes value (with length `TOTAL_INDEX_LENGTH_BYTES`).
+
+### Event Root
+
+In this LIP we introduce a new block header property, `eventRoot`. 
+The events root at a certain height `h`, `eventRoot(h)`, is the root of the sparse Merkle tree computed from all key-value pairs of all events emitted from the block at height `h` (the events tree). 
+A block header without any events has only one empty node in the events tree therefore the event root equals the `EMPTY_HASH`.
+
+### Standard Events
+
+Each module defines a standard event indicating the result of a transaction processing, with the following properties:
+
+* `moduleID`: The module ID of the transaction.
+* `typeID`: The constant `EVENT_TYPE_ID_STANDARD`.
+* `topics`: An array containing the transaction ID.
+* `data`: A boolean value set to `true` if the transaction was processed successfully, `false` otherwise. The event `data` JSON schema is given below.
```

Comment:
```suggestion
* `data`: A Boolean value set to `true` if the transaction was processed successfully, `false` otherwise. The event `data` JSON schema is given below.
```